### PR TITLE
Keep kube-apiserver HPA scale down mode `Auto` even when scale down is disabled.

### PR DESCRIPTION
**Pull request by amshuman-kr**
**How to categorize this PR?**

/area control-plane auto-scaling
/kind bug

**What this PR does / why we need it**:

Keep kube-apiserver HPA scale down mode `Auto` even when scale down is disabled. The scale down is naturally disabled because `minReplicas` and `maxReplicas` are set to be equal.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

HVPA [deletes the HPA resource ](https://github.com/gardener/hvpa-controller/blob/531ed4269a58f51bc7e5ab2ca17d0871cd8a6e53/controllers/hvpa_controller.go#L386-L388) if the [HPA scale down mode is switched off](https://github.com/gardener/hvpa-controller/blob/531ed4269a58f51bc7e5ab2ca17d0871cd8a6e53/controllers/hvpa_controller.go#L384) because HVPA v1 has no good way to control how the HPA recommendations are applied (in [HVPA v2](https://github.com/gardener/hvpa-controller/pull/76), [there is a possibility](https://github.com/gardener/hvpa-controller/blob/master/docs/proposals/HVPA-EP-0001/HVPA-EP-001.md#unpredictable-scaling-behaviour-with-0--vpaweight--100) to control this).

So, when the shoot [control-plane scale down is disabled](https://github.com/gardener/gardener/blob/35db6b62be458890e1c3fabb644a6d52acc435ab/pkg/apis/core/v1beta1/constants/types_constants.go#L223-L227), the HPA resource is effectively deleted. If the `kube-apiserver` was not already scaled up to `maxReplicas` when this happens, then HVPA will not pro-actively apply VPA recommendation (waiting for HPA to scale up to `maxReplicas` which will never happen).

When the shoot [control-plane scale down is disabled](https://github.com/gardener/gardener/blob/35db6b62be458890e1c3fabb644a6d52acc435ab/pkg/apis/core/v1beta1/constants/types_constants.go#L223-L227), HPA `minReplicas` and `maxReplicas` are already [set to be the same](https://github.com/gardener/gardener/blob/35db6b62be458890e1c3fabb644a6d52acc435ab/pkg/operation/botanist/kubeapiserver.go#L64-L67), which effectively disables scale down. So, this PR reverts the HPA scale down mode back to `Auto` to make sure the HPA resource (with `minReplicas` and `maxReplicas` same) remains in place.

An alternative would have been to make sure the `kube-apiserver` `Deployment` is scaled to `maxReplicas` but that change seems to require further refactoring in the codebase.

**Release note**:

```bugfix operator
Keep kube-apiserver HPA scale down mode `Auto` even when scale down is disabled. The scale down is naturally disabled because `minReplicas` and `maxReplicas` are set to be equal.
```

### CODE REVIEW

File: pkg/operation/botanist/component/kubeapiserver/hvpa.go [hunk 1]
```diff
@@ -125,7 +125,7 @@ func (k *kubeAPIServer) reconcileHVPA(ctx context.Context, hvpa *hvpav1alpha1.Hv
 			},
 			ScaleDown: hvpav1alpha1.ScaleType{
 				UpdatePolicy: hvpav1alpha1.UpdatePolicy{
-					UpdateMode: &scaleDownUpdateMode,
+					UpdateMode: &updateModeAuto,
```
Comment (vpnachev) [inline pkg/operation/botanist/component/kubeapiserver/hvpa.go]: Please add a short comment (few words) that HPA does not work with mode `Off` and needs to be always `Auto` even though scale down is disabled.

**Review by vpnachev — CHANGES_REQUESTED** [review]

File: pkg/operation/botanist/component/kubeapiserver/hvpa.go (hunk 1 above)
Comment (amshuman-kr) [reply pkg/operation/botanist/component/kubeapiserver/hvpa.go]: Done in https://github.com/gardener/gardener/commit/aff93c9d6f639aa582efa44d61826dadffb86b8e

**Comment by amshuman-kr** [comment]
@vpnachev Thanks for the review. I have added the comment to describe the situation as you suggested. Can you PTAL?

File: pkg/operation/botanist/component/kubeapiserver/hvpa.go [hunk 2]
```diff
@@ -125,7 +125,12 @@ func (k *kubeAPIServer) reconcileHVPA(ctx context.Context, hvpa *hvpav1alpha1.Hv
 			},
 			ScaleDown: hvpav1alpha1.ScaleType{
 				UpdatePolicy: hvpav1alpha1.UpdatePolicy{
-					UpdateMode: &scaleDownUpdateMode,
+					/*
+					 * HPA update mode needs to be always `Auto` because it is not supported in HPA.
+					 * To disable HPA scale down it is sufficient to set HPA minReplicas and
+					 * maxReplicas to be the same.
+					 */
```
Comment (vpnachev) [inline pkg/operation/botanist/component/kubeapiserver/hvpa.go]:
```suggestion
```

File: pkg/operation/botanist/component/kubeapiserver/hvpa.go (hunk 1 above)
Comment (vpnachev) [reply pkg/operation/botanist/component/kubeapiserver/hvpa.go]:
```suggestion
					UpdateMode: &updateModeAuto, # HPA does not work with update mode 'Off' and needs to be always 'Auto' even though scale down is disabled.
```

**Review by vpnachev — CHANGES_REQUESTED** [review]

**Comment by amshuman-kr** [comment]
@vpnachev I have applied your suggestions. Can you PTAL?

**Comment by amshuman-kr** [comment]
I am not sure why `make generate` is failing in the build. It works locally.

**Review by vpnachev — APPROVED** [review]
/lgtm

**Comment by amshuman-kr** [comment]
I have cherry picked to `release-v1.28` in #4466. Automatic cherry-pick to `release-v1.27` and `release-v1.26` is failing because of conflicts. I will manually cherry-pick later today.

**Comment by amshuman-kr** [comment]
Cherry-picked to `release-v1.27` in #4468 and on `release-v1.26` in #4469.

**Pull request merged**